Jgfouca/port p3 funcs

jgfouca opened the pull request:
Fixes #34

bartgol commented on components/scream/src/physics/p3/p3_functions.hpp:
```diff
@@ -0,0 +1,168 @@
+    const ko::Unmanaged<view_1d<const Spack> >& inv_rho, // 1/rho
+    const ko::Unmanaged<view_1d<const Spack> >& inv_dzq,
+    const MemberType& team,
+    const Int& nk, const Int& k_bot, const Int& k_top, const Int& kdir, const Scalar& dt_sub,
```
I see kdir a lot in the routines, and it appears to be used to switch between two different scenarios. Yet, I have no idea what it represent. Is there a more self-explanatory name for this variable? If not (e.g., it would be too long), can we add a quick explanation in the comment above (where some inputs/outputs are explained)?

jgfouca replied:
It represents the directions that columns are processed (bottom-up vs top-down). I'll add some comments.

AaronDonahue replied:
It basically enforces the direction for all the loops to go in the appropriate direction instead of from 1 to # of levels.  It's important for things like sedimentation, which should go from top to bottom.

bartgol commented on components/scream/src/physics/p3/p3_functions_find_impl.hpp:
```diff
@@ -0,0 +1,97 @@
+  } else {
+    if (kdir == -1) {
+      Kokkos::parallel_reduce(
+        Kokkos::TeamThreadRange(team, kbot - ktop + 1), [&] (Int k_, int& lmax) {
```
We are used to see either all 3 levels of parallelism (down to ThreadVectorRange) or some variation of Kokkos::single. Concerns are, obviously, for GPU builds. In this case, it's fine because there are no arithmetic operations, and the only operation happening is a min, which is a sort of 'fixed point'. It would be worth it to add a comment, so the user does not get puzzled about possible race conditions.

ambrad replied:
We'll have to come up with a general pattern for communicating things like this. One possibility is to use `scream_kassert_msg` in every function that has expectations about Kokkos vector size.

Separately, and for clarification w.r.t. these comments, P3 always sets the Kokkos vector size to 1; all ||ism (for 128 levels, 128 threads = 4 warps).

bartgol replied:
Ah ok. I was thinking about it, but I wasn't sure. That's another good solution; asserts are always verbose enough in explaining the culprit.

bartgol commented on components/scream/src/physics/p3/p3_functions_upwind_impl.hpp:
```diff
@@ -0,0 +1,140 @@
+
+  // calculate fluxes
+  Kokkos::parallel_for(
+    Kokkos::TeamThreadRange(team, kmax - kmin), [&] (Int k_) {
```
Again, I think it's worth to specify this is ok on GPU. E.g., something like "On GPU, these updates will be executed by all threads in the team vector length. That's fine, they are all simply going to write the same number. No race condition."

ambrad commented:
If you run `make baseline` in master and then rebuild the branch and run `ctest -R p3_regression`, does it pass? For me it's passing in master but failing in this branch.

ambrad commented:
I bet the baseline diff comes from a change in constants. I'll see if I can isolate the changes and confirm.

ambrad commented:
Confirmed. Fixed by this diff:
```
diff --git a/components/scream/src/physics/p3/p3_ic_cases.cpp b/components/scream/src/physics/p3/p3_ic_cases.cpp
index 32f0b2a..02f2b7c 100644
--- a/components/scream/src/physics/p3/p3_ic_cases.cpp
+++ b/components/scream/src/physics/p3/p3_ic_cases.cpp
@@ -47,10 +47,15 @@ FortranData::Ptr make_mixed () {
   // To get potential temperature, start by making absolute temperature vary
   // between 150K at top of atmos and 300k at surface, then convert to potential
   // temp.
+  static constexpr double
+    Rd = 287,      // gas constant for dry air, J/kg/K
+    cp = 1004,     // heat constant of air at constant pressure, J/kg
+    p0 = 100000;   // reference pressure, Pa
   FortranData::Array1 T("T", nk);
   for (k = 0; k < nk; ++k) {
     T(k) = 150 + 150/double(nk)*k;
-    d.th(0,k) = T(k)*std::pow(Real(Constants<double>::P0/d.pres(0,k)), Real(Constants<double>::RD/Constants<double>
+    d.th(0,k) = T(k)*std::pow(Real(p0/d.pres(0,k)), Real(Rd/cp));
+    //d.th(0,k) = T(k)*std::pow(Real(Constants<double>::P0/d.pres(0,k)), Real(Constants<double>::RD/Constants<doubl
   }
 
   // The next section modifies inout variables to satisfy weird conditions
@@ -93,7 +98,8 @@ FortranData::Ptr make_mixed () {
       d.pres(0,nk-1) + 0.5*(d.pres(0,nk-1) - d.pres(0,nk-2))/(1 - 0) :
       0.5*(d.pres(0,k) + d.pres(0,k+1));
     const auto dpres = phi - plo;
-    d.dzq(0,k) = Constants<double>::RD*T(k)/(g*d.pres(0,k))*dpres;
+    d.dzq(0,k) = Rd*T(k)/(g*d.pres(0,k))*dpres;
+    //d.dzq(0,k) = Constants<double>::RD*T(k)/(g*d.pres(0,k))*dpres;
   }
```
So this PR is not BFB, but this diff explains the diff, and it's OK.

jgfouca commented on components/scream/src/physics/p3/p3_constants.hpp:
```diff
@@ -0,0 +1,63 @@
+{
+  static constexpr Scalar INV_RHOW = 1.e-3;
+  static constexpr Scalar RHOW     = 997.0;
+  static constexpr Scalar THRD     = 1.0/3.0;
```
Most of these constants came from the fortran p3 implementation. I agree, some of the names are a bit weak.

> Besides, is there a reason why we have a name for 1/3, rather than writing `blah/3.0` in the code?

I think this is another fortranism.

ambrad commented on components/scream/src/physics/p3/p3_constants.hpp:
```diff
@@ -0,0 +1,63 @@
+  static constexpr Scalar RHOSUR   = P0/(RD*273.15);
+  static constexpr Scalar CP       = 1005.0;          // heat constant of air at constant pressure, J/kg
+  static constexpr Scalar INV_CP   = 1.0/CP;
+};
```
I'd like to advocate for not using all-caps for these constants. Physical constants often have lower, upper, or mixed case by convention. For example, RD makes more sense as Rd or R_d. PI should be pi. These don't have to change all at once, but for the ones that are used in the Fortran P3, I think we should follow that convention. @bartgol, @AaronDonahue, thoughts?

More generally, since C++ has namespaces (via either `namespace` or, as in this case, structs and classes), we don't have to use special casing to indicate something is a global constant; the struct name `Constants` fulfills that role in this case.

jgfouca commented on components/scream/src/physics/p3/p3_constants.hpp:
```diff
@@ -0,0 +1,63 @@
+template <typename Scalar>
+struct Constants
+{
+  static constexpr Scalar INV_RHOW = 1.e-3;
```
Not sure, it was like this in fortran p3.

jgfouca replied:
I agree. What I meant to say is that I don't know what the intent is behind this variable since I'm not an expert in the original p3 impl. Here's where it was being used:

```
prt_liq(i) = prt_liq(i) + prt_accum*INV_RHOW*odt
```

AaronDonahue replied:
*Brought over from SLACK* 

`prt_liq` is used by the mass checker to figure out how much mass has left the system via the surface precip
`prt_accum` is the sedimentation at the surface

AaronDonahue replied:
per @bartgol comment: In the implementation that will eventually make it into SCREAM we have fixed this inconsistentcy.  In short, yes inv_rhow should not be defined this way, instead it should be 1.0/rhow which we have done

ambrad replied:
@AaronDonahue is there a reason for INV_RHOW to not equal 1.0/RHOW?

AaronDonahue replied:
Not really.  I think the new implementation of P3 that Hugh and his team just recently released fixes this to be 1.0/RHOW.

AaronDonahue replied:
``` cp     = 1005.
 inv_cp = 1./cp
 g      = 9.816
 rd     = 287.15
 rv     = 461.51
 ep_2   = 0.622
 rhosur = 100000./(rd*273.15)
 rhosui = 60000./(rd*253.15)
 ar     = 841.99667
 br     = 0.8
 f1r    = 0.78
 f2r    = 0.32
 ecr    = 1.
 rhow   = 1000.
 cpw    = 4218.
 inv_rhow = 1./rhow  !inverse of (max.) density of liquid water
 mu_r_constant = 1.  !fixed shape parameter for mu_r```

ambrad replied:
@jgfouca then let's break BFB w.r.t. master with this change, too: define INV_RHOW as 1.0/RHOW.

AaronDonahue replied:
Will it make much difference to your effort when we bring in the P3 branch which will have a set of changes to this set of constants?  Would it be helpful for me to share the constants we plan on using in the final version as they are defined now, or would switching over be just as easy when we get to the point that we are bringing in P3?

AaronDonahue replied:
for example, many of these constants will be changed to adopt the E3SM set of universal constants.

ambrad replied:
@AaronDonahue, I think it's OK to leave the changes to the constants for later. This PR focuses on mostly low-level code. The issue of constants came up only b/c they affected the ICs for the current P3 baseline regression test.

jgfouca commented on components/scream/src/physics/p3/p3_functions.hpp:
```diff
@@ -0,0 +1,168 @@
+  using Scalar = ScalarT;
+  using Device = DeviceT;
+
+  template <typename S> using BigPack = scream::pack::BigPack<S>;
```
Agree.

jgfouca commented:
> So this PR is not BFB, but this diff explains the diff, and it's OK.

Apologies. When I was writing this code, I intended to make a github comment regarding these constants, but I forgot. In particular, why "cp" was set to 1004 in the original code but is 1005 in the micro app?

AaronDonahue commented:
*Brought over from SLACK* 

`prt_liq` is used by the mass checker to figure out how much mass has left the system via the surface precip
`prt_accum` is the sedimentation at the surface

jgfouca commented:
There has been a lot of traffic on this PR this morning. I think I have addressed all the feedback with the exceptions of:
1) Clarifying race-condition safety of our kernels
2) Potential refactor/renaming of constants

For (1) should I just assert that kokkos vector size is 1 at the beginning of every p3_function that has kokkos dispatches?

I'm tempted to punt on (2) for now.

ambrad commented:
@bartgol should weigh in on 1. IMO, for P3-specific functions, we should _not_ assert in every function, as that gets verbose, and all of P3 makes that assumption. Thus, IMO, we'll place one assumption in the top-level spot when that time comes.

I'm OK with 2. We'll talk with others on tmw's telecon about it.

bartgol commented:
How many functions will p3 consist of, that would/could require such checks? If we are talking about <10, then I vote for an explicit assert (quite verbose, does not penalize performance). If that number increases, I guess we can just put assert in top level functions, and possibly also a comment per header, so that the user (which _should_ read the comments) can be aware of that, even if he/she does not look at the top level routines...

ambrad commented:
Good question.

First, keep in mind the user will call a top-level routine that does the kernel dispatch(es).

Second, b/c it's good practice for many reasons, we'll have a lot of functions. It's at core ~1700 lines of Fortran, so I estimate ~30 functions at minimum.

Third, general API-level functions will go in share/, not p3/, and these would of course state assumptions through compile-time or runtime checks. But p3-specific functions IMO don't have to offer the same interface checking that an API-type function does. We can expect quite a lot of details to have to hold when calling a p3-specific function; Kokkos team vector size is just one.

I don't normally argue against _more_ checking, but here I think the threat is boilerplate. Better to thoughtfully check things that really might go wrong w/in P3, given the context, than to boilerplate-check in every function for Kokkos parameters. If we start doing that, then we would also logically check that every array is compatible, etc. In reality, only API functions do that, the idea being that you don't end up in an internal impl function unless you've already passed through an API-level check.

bartgol commented:
Sounds reasonable. But I'd still argue for putting a comment at the beginning of the file explaining all the 'not so common' assumptions used throughout the implementation. By 'not so common' I mean something like the vector length, not something like 'the pointers should not be null'...

It's true that you 'usually' pass through the API layers, but a bit of verbosity does not hurt. As I said, some comments at the beginning of the file, or maybe in a 'general' p3 header, would do. Even better, a doc file in some folder containing 'info you should read before looking at the code'.

Anyhow, I don't want to be pedantic. For now we can simply put checks in top-level routines. Later we can discuss if/how to be more verbose.
